Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -5675,21 +5661,20 @@ bool WebGLRenderingContextBase::validateArrayBufferType(const char* functionName
     return true;
 }
 
-bool WebGLRenderingContextBase::validateTexFuncData(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset)
+bool WebGLRenderingContextBase::validateTexFuncData(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset, size_t* totalBytesRead)
 {
     // All calling functions check isContextLostOrPending, so a duplicate check is not
     // needed here.
-    if (!pixels) {
+    if (!pixels && disposition != NullAllowed) {
         ASSERT(disposition != NullNotReachable);
-        if (disposition == NullAllowed)
-            return true;
         synthesizeGLError(GraphicsContextGL::INVALID_VALUE, functionName, "no pixels");
         return false;
     }
 
-    if (!validateSettableTexInternalFormat(functionName, format))
+    if (pixels && !validateSettableTexInternalFormat(functionName, format))
```

Comment:
The settable tex internal format validation must happen regardless of whether `pixels` is passed or not. This is part of the bug about historical early return that you are fixing.

Reply:
After scrutinising the use of `validateSettableTexInernalFormat`, I've concluded that the intention is only to call this function when the API is passed image data. This is to reject `gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, new Uint16Array(1))` but accept `gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, null)` as specified by `WEBGL_depth_texture` extension.

`validateTexFuncFormatAndType` validates the combination of `internalformat`, `format`, and `type` before reaching this point.